### PR TITLE
feat: make agent stats' cardinality configurable

### PR DESCRIPTION
**Pull request by dannykopping**
Implements https://github.com/coder/coder/issues/12221

When stats from agents are collected, they are aggregated by 4 dimensions: `agent_name`, `template_name`, `workspace_name`, and `username`. This can result in some very high cardinality metrics being scraped by Prometheus in large environments.

This PR adds the ability to tune which labels are include in this aggregation, therefore reducing the cardinality.

For example:

```prom
agent_sessions_total{agent_name="main",magic_type="ssh",pty="yes",template_name="docker",username="danny",workspace_name="test1"} 6
agent_sessions_total{agent_name="main",magic_type="ssh",pty="yes",template_name="docker",username="danny",workspace_name="test2"} 5
```

With hundreds of active workspaces, each having a unique name, the cardinality may be unacceptable. In this case the operator may choose to configure `CODER_PROMETHEUS_AGGREGATE_AGENT_STATS_BY=username` to rather aggregate by user, summing all the metrics' values and only producing a single metric series:

```prom
agent_sessions_total{magic_type="ssh",pty="yes",username="danny"} 11
```

Multiple labels can be provided, e.g. `CODER_PROMETHEUS_AGGREGATE_AGENT_STATS_BY=username,agent_name`

```prom
agent_sessions_total{agent_name="main",magic_type="ssh",pty="yes",username="danny"} 11
```

**The current behaviour remains the default; if no value is passed to `CODER_PROMETHEUS_AGGREGATE_AGENT_STATS_BY` there will be no change.**

_Note to reviewer: I made a sweeping refactor to define the label names in a single location instead of duplicating them, which may make the PR larger than it seems._

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -26,6 +26,7 @@
           bat
           cairo
           curl
+          delve
```
Comment (dannykopping) [inline flake.nix:29]: Drive-by fix to get `scripts/develop.sh --debug` working

File: coderd/prometheusmetrics/aggregator.go [hunk 2]
```diff
@@ -8,8 +8,11 @@ import (
 	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
 	"golang.org/x/xerrors"
 
+	"github.com/coder/coder/v2/coderd/agentmetrics"
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:14]: Unfortunately our formatter doesn't handle merging import groups and leaves things in a messy state (depending on what program injected them). 😔

If you notice these, please feel free to fix, but the standard is we try our best but sometimes these slip through, so don't worry too much.

File: coderd/prometheusmetrics/aggregator.go [hunk 3]
```diff
@@ -102,13 +108,28 @@ func hashKey(req *updateRequest, m *agentproto.Stats_Metric) metricKey {
...
 func (am *annotatedMetric) asPrometheus() (prometheus.Metric, error) {
-	labels := make([]string, 0, len(agentMetricsLabels)+len(am.Labels))
-	labelValues := make([]string, 0, len(agentMetricsLabels)+len(am.Labels))
+	var (
+		baseLabelNames  = am.aggregateByLabels
+		baseLabelValues []string
+		extraLabels     = am.Labels
+	)
+
+	for _, label := range am.aggregateByLabels {
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:117]: Perhaps this was the intent here?

```suggestion
	for _, label := range baseLabelNames {
```

File: coderd/prometheusmetrics/aggregator.go [hunk 4]
```diff
@@ -118,10 +139,48 @@ func (am *annotatedMetric) asPrometheus() (prometheus.Metric, error) {
...
+	case agentmetrics.UsernameLabel:
+		labelVal = am.username
+	default:
+		return "", xerrors.Errorf("unexpected label: %q", label)
+	}
+
+	return labelVal, nil
+}
+
+func (am *annotatedMetric) clone() annotatedMetric {
```
Comment (mafredri) [inline coderd/prometheusmetrics/aggregator.go:165]: Would `shallowCopy` be more appropriate as a name, or maybe `cloneShallow`? I'm thinking it would be clearer since we're not cloning slices. Alt: should this be a deep clone?

File: codersdk/deployment.go [hunk 5]
```diff
@@ -942,6 +953,33 @@ when required by your organization's security policy.`,
...
+			Name:        "Prometheus Aggregate Agent Stats By",
+			Description: fmt.Sprintf("When collecting agent stats, aggregate metrics by a given set of comma-separated labels to reduce cardinality. Accepted values are %q.", AcceptedMetricAggregationLabels),
+			Flag:        "prometheus-aggregate-agent-stats-by",
+			Env:         "CODER_PROMETHEUS_AGGREGATE_AGENT_STATS_BY",
+			Value: clibase.Validate(&c.Prometheus.AggregateAgentStatsBy, func(value *clibase.StringArray) error {
+				if value == nil {
+					return nil
+				}
+
+				acceptable := make(map[string]any, len(AcceptedMetricAggregationLabels))
```
Comment (mafredri) [inline codersdk/deployment.go:966]: I think using `map[string]bool` would be preferable/clearer here, also simplifies the map lookup later. Typically I'd use either `bool` or if we're looking for space savings, I'd use the zero struct (`map[string]struct{}`).

**Review by mafredri — COMMENTED** [review]
Noticed some minor nits, but in general look good, nice work!

File: coderd/agentmetrics/labels.go [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package agentmetrics
+
+const (
+	TemplateNameLabel  = "template_name"
```
Comment (mtojek) [inline coderd/agentmetrics/labels.go:4]: nit: sort alphabetically?

File: coderd/prometheusmetrics/aggregator.go [hunk 7]
```diff
@@ -216,15 +332,41 @@ func (ma *MetricsAggregator) Run(ctx context.Context) func() {
 			case outputCh := <-ma.collectCh:
 				ma.log.Debug(ctx, "collect metrics")
 
+				var input []annotatedMetric
 				output := make([]prometheus.Metric, 0, len(ma.store))
-				for _, m := range ma.store {
+
+				// If custom aggregation labels have not been chosen, generate Prometheus metrics without any pre-aggregation.
+				// This results in higher cardinality, but may be desirable in larger deployments.
+				// Default behavior.
```
Comment (mtojek) [inline coderd/prometheusmetrics/aggregator.go:340]: nit: maybe add an empty line before this comment. Otherwise, it looks like a single paragraph.

File: codersdk/deployment.go [hunk 8]
```diff
@@ -55,6 +58,11 @@ const (
 	FeatureControlSharedPorts         FeatureName = "control_shared_ports"
 )
 
+var AcceptedMetricAggregationLabels = []string{
+	agentmetrics.TemplateNameLabel, agentmetrics.AgentNameLabel,
+	agentmetrics.UsernameLabel, agentmetrics.WorkspaceNameLabel,
+}
```
Comment (johnstcn) [inline codersdk/deployment.go:64]: suggestion: expose `agentmetrics.LabelAll []string` and then just import that here; if we end up adding more labels here it's more likely that we'll forget to handle it here due to the non-locality.

File: codersdk/deployment.go [hunk 9]
```diff
@@ -942,6 +951,33 @@ when required by your organization's security policy.`,
 			Group:       &deploymentGroupIntrospectionPrometheus,
 			YAML:        "collect_agent_stats",
 		},
+		{
+			Name:        "Prometheus Aggregate Agent Stats By",
+			Description: fmt.Sprintf("When collecting agent stats, aggregate metrics by a given set of comma-separated labels to reduce cardinality. Accepted values are %s.", strings.Join(AcceptedMetricAggregationLabels, ", ")),
+			Flag:        "prometheus-aggregate-agent-stats-by",
+			Env:         "CODER_PROMETHEUS_AGGREGATE_AGENT_STATS_BY",
+			Value: clibase.Validate(&c.Prometheus.AggregateAgentStatsBy, func(value *clibase.StringArray) error {
```
Comment (mtojek) [inline codersdk/deployment.go:959]: Would it be beneficial to extract this function outside and have it tested?

File: codersdk/deployment.go [hunk 10]
```diff
@@ -942,6 +951,33 @@ when required by your organization's security policy.`,
...
+				if value == nil {
+					return nil
+				}
+
+				acceptable := make(map[string]any, len(AcceptedMetricAggregationLabels))
+				for _, label := range AcceptedMetricAggregationLabels {
+					acceptable[label] = nil
```
Comment (mtojek) [inline codersdk/deployment.go:966]: should we count the occurences? in other words, is "template,username,template" a valid array?

File: coderd/prometheusmetrics/prometheusmetrics.go [hunk 11]
```diff
@@ -335,11 +329,17 @@ func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Regis
 	}, nil
 }
 
-func AgentStats(ctx context.Context, logger slog.Logger, registerer prometheus.Registerer, db database.Store, initialCreateAfter time.Time, duration time.Duration) (func(), error) {
+var DefaultAgentStatsLabels = []string{agentmetrics.AgentNameLabel, agentmetrics.UsernameLabel, agentmetrics.WorkspaceNameLabel}
```
Comment (mtojek) [inline coderd/prometheusmetrics/prometheusmetrics.go:332]: nit: maybe we should move to `agentmetrics` 🤔

File: coderd/prometheusmetrics/aggregator.go [hunk 12]
```diff
@@ -174,9 +233,66 @@ func NewMetricsAggregator(logger slog.Logger, registerer prometheus.Registerer,
...
+}
+
+func newLabelAggregator(size int) *labelAggregator {
+	return &labelAggregator{
+		aggregations: make(map[string]float64, size),
+		metrics:      make(map[string]annotatedMetric, size),
+	}
+}
+
+func (a *labelAggregator) aggregate(am annotatedMetric, labels []string) error {
```
Comment (mtojek) [inline coderd/prometheusmetrics/aggregator.go:254]: QQ: What is the extra overhead if `labelAggregator` is used together with default set of labels?

I'm also curious if this can be implemented using a strategy pattern, aggregated vs. non-aggregated. It is not a mandatory requirement, rather a challenge. I'm afraid that messing up with the current aggregator may result in unexpected events/errors.

File: coderd/prometheusmetrics/aggregator_test.go [hunk 13]
```diff
@@ -291,18 +297,322 @@ func TestUpdateMetrics_MetricsExpire(t *testing.T) {
 	}, testutil.WaitShort, testutil.IntervalFast)
 }
 
+func TestLabelsAggregation(t *testing.T) {
```
Comment (johnstcn) [inline coderd/prometheusmetrics/aggregator_test.go:300]: Nice test! 👍

File: codersdk/deployment.go (hunk 10 above)
Comment (johnstcn) [reply codersdk/deployment.go:966]: I wouldn't say that it's invalid enough to spit out an error tbh.

File: codersdk/deployment.go (hunk 9 above)
Comment (johnstcn) [reply codersdk/deployment.go:959]: 👍 probably no harm to do this

File: codersdk/deployment.go (hunk 10 above)
Comment (mafredri) [reply codersdk/deployment.go:966]: I agree with @johnstcn, at most a notice that we're ignoring the duplicate labels but not sure if it's worth the effort.

File: coderd/agentmetrics/labels.go (hunk 6 above)
Comment (mafredri) [reply coderd/agentmetrics/labels.go:4]: Autocomplete is pretty good at fuzzy matching these days. That said, there is value in both approaches, I often struggle to decide which to go with 😄.

File: coderd/prometheusmetrics/aggregator.go (hunk 12 above)
Comment (dannykopping) [reply coderd/prometheusmetrics/aggregator.go:254]: > QQ: What is the extra overhead if labelAggregator is used together with default set of labels?

```go
goos: darwin
goarch: arm64
pkg: github.com/coder/coder/v2/coderd/prometheusmetrics
                          │ /tmp/no-aggregation.out │         /tmp/aggregation.out         │
                          │         sec/op          │    sec/op     vs base                │
_MetricsAggregator_Run-12              2.219m ± 34%   3.124m ± 31%  +40.77% (p=0.011 n=10)

                          │ /tmp/no-aggregation.out │         /tmp/aggregation.out         │
                          │          B/op           │     B/op      vs base                │
_MetricsAggregator_Run-12              2.800Mi ± 1%   3.288Mi ± 4%  +17.44% (p=0.000 n=10)

                          │ /tmp/no-aggregation.out │        /tmp/aggregation.out         │
                          │        allocs/op        │  allocs/op   vs base                │
_MetricsAggregator_Run-12               36.89k ± 0%   43.17k ± 0%  +17.03% (p=0.000 n=10)
```

Seems like quite a lot of overhead because of all the extra allocations: the profile indicates about 30% of time spent on GC, which may not be realistic in a normal running process but it's still indicative. I don't think it's too much of an issue unless we're aggregating metrics from hundreds or thousands of agents. What do you think?

> I'm also curious if this can be implemented using a strategy pattern, aggregated vs. non-aggregated. It is not a mandatory requirement, rather a challenge.

It'd certainly make it cleaner, although I doubt it's worth the time investment. What do you think?

> I'm afraid that messing up with the current aggregator may result in unexpected events/errors.

My changes don't alter the behaviour of the current aggregator since we're taking a copy of the `annotationMetric`s.

File: codersdk/deployment.go (hunk 10 above)
Comment (dannykopping) [reply codersdk/deployment.go:961]: Agreed; I think we can leave this for now because duplicates are filtered out and it doesn't alter behaviour at all.

**Review by mafredri — APPROVED** [review]
LGTM! I know tests aren't passing now but looks unrelated so I don't need to re-review, nice work!

File: coderd/prometheusmetrics/aggregator.go (hunk 12 above)
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go:254]: Oh, I didn't expect the deep investigation :)

> It'd certainly make it cleaner, although I doubt it's worth the time investment. What do you think?

This is a question about our confidence in this change. If it seems to be stable/not loosing metrics, I guess we can risk and leave it as the only/default implementation 👍 

Anyway, feel free to resolve this comment. Thanks!

File: codersdk/deployment.go (hunk 10 above)
Comment (mtojek) [reply codersdk/deployment.go]: Ok

**Review by mtojek — APPROVED** [review]
Nice!

> I know tests aren't passing now but looks unrelated so I don't need to re-review, nice work!

Agree, feel free to `t.Skip` them for now.

**Review by johnstcn — APPROVED** [review]
